Document System.Text.RegularExpressions.GroupCollection

carlossanlop opened the pull request:
This one was interesting because the class implements IList and ICollection, which have shared method names.
Strangely, there were a couple of members that were duplicate, so I removed them. Please let me know if it was wrong to do so.

Here is the code: [GroupCollection.cs](https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/GroupCollection.cs)

I based the documentation on the interface comments:
- ICollection: https://docs.microsoft.com/en-us/dotnet/api/system.collections.icollection?view=netcore-3.0
- IList: https://docs.microsoft.com/en-us/dotnet/api/system.collections.ilist?view=netcore-3.0

Area owners: @ViktorHofer, @maryamariyan
Language review: @rpetrusha, @Jemash

carlossanlop commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -971,115 +1000,10 @@
-          <paramref name="arrayIndex" /> is outside the bounds of <paramref name="array" />.-or-<paramref name="arrayIndex" /> plus <see cref="P:System.Text.RegularExpressions.GroupCollection.Count" /> is outside the bounds of <paramref name="array" />.</exception>
-      </Docs>
-    </Member>
-    <Member MemberName="System.Collections.ICollection.IsSynchronized">
```
This one I'm 100% sure was correct to remove because it's already documented in l[ine 358](https://github.com/dotnet/dotnet-api-docs/pull/2803/files#diff-69e2ff9cb3ca19ddcb83db53484b00ceR358). The code file only has one method version.

carlossanlop replied:
Restoring.

carlossanlop commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -971,115 +1000,10 @@
-        <remarks>To be added.</remarks>
-      </Docs>
-    </Member>
-    <Member MemberName="System.Collections.ICollection.SyncRoot">
```
`SyncRoot` I'm also 100% sure was correct to remove because it's already documented in [line 593](https://github.com/dotnet/dotnet-api-docs/pull/2803/files#diff-69e2ff9cb3ca19ddcb83db53484b00ceR593). and the code file only has one method version.

carlossanlop replied:
Restoring.

carlossanlop commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -971,115 +1000,10 @@
-        <remarks>To be added.</remarks>
-      </Docs>
-    </Member>
-    <Member MemberName="System.Collections.ICollection.CopyTo">
```
I *think* it was correct to remove this one because there are already 2 implementations documented in this file (lines [112 ](https://github.com/dotnet/dotnet-api-docs/pull/2803/files#diff-69e2ff9cb3ca19ddcb83db53484b00ceR112) and [186](https://github.com/dotnet/dotnet-api-docs/pull/2803/files#diff-69e2ff9cb3ca19ddcb83db53484b00ceR186)), one for an `Array` as first parameter and another one for a `Group[]` as first parameter.

This third version may have been mistakenly added by mdoc back in January. Notice the MemberName.

carlossanlop replied:
Ok thank you for letting me know. I'll restore them.

maryamariyan commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -103,9 +103,9 @@
-        <returns>To be added.</returns>
+        <param name="key">A string with the name of the captured group to locate.</param>
+        <summary>Determines whether the group collection contains a captured group identified by the specified name.</summary>
+        <returns><see langword="true"/> if the groupcollection contains a captured group identified by the specified key; <see langword="false" /> otherwise.</returns>
```
is it ok for groupcollection to be lowercase like this?

carlossanlop replied:
Nope, my mistake. Fixing it:

```suggestion
        <returns><see langword="true"/> if the group collection contains a captured group identified by the specified key; <see langword="false" /> otherwise.</returns>
```

maryamariyan commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -1351,9 +1278,11 @@
-        <summary>To be added.</summary>
+        <param name="item">This method is not supported.</param>
+        <summary>Implementation from <see cref="T:System.Collections.Generic.IList" />. This method is not supported.</summary>
+        <returns>This method is not supported.</returns>
```
`<returns>This method is not supported.</returns>`

Isn't it better to say this method throws with NotSupportedException or something?

maryamariyan replied:
I found two other places where returns tag says "This method is not supported."

rpetrusha commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -1416,10 +1346,10 @@
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="key">A string with the group name key to look for.</param>
+        <param name="value">The out parameter where the retrieved group will be saved if found; <see langword="null" /> if not found.</param>
```
```suggestion
        <param name="value">When the method returns, the group whose name is <paramref name="key" />, if it is found; otherwise, <see langword="null" /> if not found.</param>
```

rpetrusha commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -1416,10 +1346,10 @@
+        <param name="key">A string with the group name key to look for.</param>
+        <param name="value">The out parameter where the retrieved group will be saved if found; <see langword="null" /> if not found.</param>
+        <summary>Attempts to retrieve a group identified by the provided name key, if it exists in the group collection.</summary>
+        <returns><see langword="true" /> if a group identified by the provided name key existed; <see langword="false" /> otherwise.</returns>
```
```suggestion
        <returns><see langword="true" /> if a group identified by the provided name key exists; <see langword="false" /> otherwise.</returns>
```

rpetrusha commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -1385,9 +1314,10 @@
       <Docs>
-        <param name="index">To be added.</param>
-        <summary>To be added.</summary>
+        <param name="index">This method is not supported.</param>
```
Similarly, use the language from previous not supported methods.

rpetrusha commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -1448,10 +1378,10 @@
-        <value>To be added.</value>
+        <summary>Gets a group enumeration with all the groups in the group collection.</summary>
+        <value>A group enumeration.</value>
         <remarks>To be added.</remarks>
```
Add to the Remarks that this is an EII -- you can copy and adapt the earlier text.

carlossanlop commented:
I addressed your comments, @rpetrusha. I had to to a `git push -f` so the previous commits are not showing up anymore, but I merged all your rewording suggestions, the EIIs and the "not supported" changed messages.

rpetrusha commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -1177,9 +1233,9 @@
+        <param name="value">The group to locate in the group collection.</param>
+        <summary>Determines whether the group collection contains a specific group item.</summary>
+        <returns><see langword="true" /> if the group item is found in the group collection; <see langword="false" /> otherwise.</returns>
         <remarks>To be added.</remarks>
```
```suggestion
        <remarks>  
          <format type="text/markdown"><![CDATA[  

This member is an explicit interface member implementation. It can be used only when the <xref:System.Text.RegularExpressions.GroupCollection> instance is cast to an <xref:System.Collections.IList> interface. 

         ]]></format>
         </remarks>
```

carlossanlop replied:
Edited your suggestion to add the `## Remarks` title.

rpetrusha commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -1212,9 +1268,9 @@
+        <param name="value">The group to locate in the group collection.</param>
+        <summary>Determines the index of a specific group in the group collection.</summary>
+        <returns>The index of the <paramref. name="item" /> if found; otherwise, -1.</returns>
         <remarks>To be added.</remarks>
```
```suggestion
This member is an explicit interface member implementation. It can be used only when the <xref:System.Text.RegularExpressions.GroupCollection> instance is cast to an <xref:System.Collections.IList> interface. ```

carlossanlop replied:
I modified your suggestion a bit to include CDATA portion, otherwise the xref won't work:

```suggestion
        <remarks> 
          <format type="text/markdown"><![CDATA[  

## Remarks

This member is an explicit interface member implementation. It can be used only when the <xref:System.Text.RegularExpressions.GroupCollection> instance is cast to an <xref:System.Collections.IList> interface. 

         ]]></format>
         </remarks>
```

rpetrusha commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -1281,8 +1338,8 @@
-        <value>To be added.</value>
+        <summary>Gets a value indicating whether the group collection has a fixed size.</summary>
+        <value><see langword="true" /> always.</value>
         <remarks>To be added.</remarks>
```
```suggestion
This member is an explicit interface member implementation. It can be used only when the <xref:System.Text.RegularExpressions.GroupCollection> instance is cast to an <xref:System.Collections.IList> interface. ```

carlossanlop replied:
Same with the CDATA:

```suggestion
         <remarks>
           <format type="text/markdown"><![CDATA[  

## Remarks

This member is an explicit interface member implementation. It can be used only when the <xref:System.Text.RegularExpressions.GroupCollection> instance is cast to an <xref:System.Collections.IList> interface. 

           ]]></format>
         </remarks>
```

rpetrusha commented on xml/System.Text.RegularExpressions/GroupCollection.xml:
```diff
@@ -1316,10 +1373,11 @@
+        <param name="index">The zero-index position of the group in the group collection.</param>
+        <summary>Gets the group in the desired position.</summary>
+        <value>The group in the desired position.</value>
         <remarks>To be added.</remarks>
```
```suggestion
This member is an explicit interface member implementation. It can be used only when the <xref:System.Text.RegularExpressions.GroupCollection> instance is cast to an <xref:System.Collections.IList> interface. ```

carlossanlop replied:
```suggestion
         <remarks>
           <format type="text/markdown"><![CDATA[  

## Remarks

This member is an explicit interface member implementation. It can be used only when the <xref:System.Text.RegularExpressions.GroupCollection> instance is cast to an <xref:System.Collections.IList> interface. 

           ]]></format>
         </remarks>
```

carlossanlop commented:
@mairaw @rpetrusha I applied your suggestions Ron, thanks for finding the additional remarks. I modified them a bit to include CDATA, otherwise the xrefs would give us trouble. Let me know if they look good while we wait for the build to finish.

rpetrusha commented:
This looks good, @carlossanlop. I just noticed that there's one more EII (ICollection.CopyTo) that could have the "This member is an EII..." text, but I'll open a PR to add it separately, and I'll merge your PR now.
